feat: flox-env-helper for grooming PATH & MANPATH

limeytexan opened the pull request:
## Proposed Changes

Reimplemented PATH and MANPATH rewriting logic by way of a new flox-env-helper script, patterned in a way after the Mac path-helper. The output of this script can be sourced at any time to set the PATH and MANPATH variables according to the current value of the FLOX_ENV_DIRS variable.

This script allowed us to remove a bunch of logic associated with the setting of PATH and MANPATH, as well as retiring the previous set-run-variables.bash script in its entirety.

My expectation is that this should close some number of open issues pertaining to version 1.3.8.

Issues resolved or helped by this PR:
* #2540 
* #2524 
* helps with #2499 
* partially addresses #2491 
* [Path issue with 1.3.8 on mac + zsh + direnv](https://discourse.flox.dev/t/path-issue-with-1-3-8-on-mac-zsh-direnv/1023/3?u=brantley)

## Release Notes

- addressed PATH and MANPATH bugs introduced with version 1.3.8

ysndr commented on assets/activation-scripts/activate.d/flox-path-helper.awk:
```diff
@@ -0,0 +1,127 @@
+#
+# So the presence of "extra" parameters in the following function
+# definition is intentional, and is used to create local variables.
+function basename(file, # local variables follow
+                  a,
+		  n) {
+  n = split(file, a, "/")
+  return a[n]
+}
+
+function remove_path_dups(path, # local variables follow
+                          path_array,
+			  dedup_array,
+			  seen_hash,
+			  i,
+			  dedup_count) {
+  dedup_count = 1
+  dedup_array[dedup_count] = ""
+  for (i = 1; i <= split(path, path_array, ":"); i++) {
+    if (seen_hash[path_array[i]] == 1) {
+      continue
+    } else {
+      dedup_array[dedup_count++] = path_array[i]
+      seen_hash[path_array[i]] = 1
+    }
+  }
+  result = dedup_array[1]
+  for (i = 2; i < dedup_count; i++) {
+    result = result ":" dedup_array[i]
+  }
+  return "\"" result "\""
+}
```
`(rand < 0.5) ? " " : "\t"`

suggestion: unless there is a specific `awk`wardness, please consider formatting this.
not usre how well that works but apparently awk has a formatter flag builtin even: https://www.gnu.org/software/gawk/manual/gawk.html#index-_002do-option

limeytexan replied:
I'll look into that, wasn't aware gawk had that feature. That said, the declaration of local variables as arguments to a function declaration is something I wasn't at all aware of, and as I was thinking of ways to make that clear this is the best I came up with. If the gawk autoformatter disagrees then I might be inclined to ignore it; we shall see.

ysndr replied:
>  That said, the declaration of local variables as arguments to a function declaration is something I wasn't at all aware of, and as I was thinking of ways to make that clear this is the best I came up with.

There got to be a different way to mark things rather than by the use of tabs or spaces in front of them, prefixing them with `_` or something like that.

limeytexan replied:
The tabs were accidental, now gone. This convention of declaring local variables is unusual - I didn't know about it until working on this PR - but I definitely think it's clearer to add the comment and split them out into multiple lines to highlight the difference.

My latest commit has it split out like this which I think is clear enough while keeping the format/indentation concise:
```
function remove_path_dups(path, # local variables follow
         path_array, dedup_array, seen_hash, i, dedup_count) {
```
Let me know if you would prefer it done differently.

limeytexan replied:
I'll go ahead and resolve the conversation for now - please reply if you would prefer something different!

ysndr commented on CONTRIBUTING.md:
```diff
@@ -214,6 +214,52 @@ section refers to settings from that plugin.
+
+Enable tracing with one of the following:
+
+1. set `FLOX_TRACE` to a non-empty value
```
suggestion: I might only be crown prince of bieke shedding, but a more specific name for this (`FLOX_ACTIVATION_TRACE`) might be more appropriate, given that it is independent of flox' tracing and specific to activations.

limeytexan replied:
Agreed, was expecting this.  😄  Wanted to have that conversation.

limeytexan replied:
Renamed the variable and logged #2554 to get the trace feature merged separately.

ysndr reviewed (commented):
I dont know if these |2| changes _need_ to be lumped together, but i appreciate that its at least separate commits.
Mind that the commit messages likely wont pass CI ^

limeytexan commented:
> I dont know if these |2| changes _need_ to be lumped together, but i appreciate that its at least separate commits. Mind that the commit messages likely wont pass CI ^

Yes, thanks for the review of the draft commit. I was def expecting to revise the commit messages before bringing it out of draft. I developed both commits at the same time then spent the extra time to split them out into different commits because I wanted the history of each commit to stand separately, but I really care less whether these flow through in one or two PRs. Will try to get it finalized, passing tests and out of draft tomorrow. FTR this patch fixes the 1.3.8 PATH problems that I've been seeing in interactive use.

limeytexan commented:
Food for thought: might want to rename this to be `flox-activate-helper` because it's in a position to "help" with all env vars, shell settings, etc. Or could be `flox-shell-helper`, or drop the mention of "helper" altogether if we want to avoid comparisons with the Apple `path-helper`?!?

limeytexan commented:
Tests now passing, and several bugs (in both CLI and tests) have been spotted and fixed. Now confident with the correctness of this approach, so now it just remains to:

* refactor/squash commits
* split out bugfixes into separate PRs as/if necessary
* rename to `flox-env-helper` to open up the possibility of using it for other grooming tasks

Will try to do that tomorrow, but may not happen until the new year.  🤞

limeytexan commented:
> Food for thought: might want to rename this to be `flox-activate-helper` because it's in a position to "help" with all env vars, shell settings, etc. Or could be `flox-shell-helper`, or drop the mention of "helper" altogether if we want to avoid comparisons with the Apple `path-helper`?!?

Renamed `s/flox.path.helper/flox.env.helper/g`

limeytexan commented:
To discuss: relation of `FLOX_ENV_DIRS` vs. `_FLOX_ACTIVE_ENVIRONMENTS`, and which is the source of truth, particularly when mixing rust and activate script activations (e.g. as happens in the nested environments found with a build). At the moment I'm finding that all tests pass when `_FLOX_ACTIVE_ENVIRONMENTS` does _not_ match reality, and wondering if that means our test coverage is missing something or if we shouldn't a) consolidate these into a single variable, managed from the `flox-env-helper` script.

dcarley commented on cli/tests/activate.bats:
```diff
@@ -1505,6 +1505,59 @@ EOF
 
 # ---------------------------------------------------------------------------- #
 
+# N.B. removed "'flox activate' only patches PATH when already activated" test,
+# because we in fact need to patch PATH with every activation to defend against
+# userShell "dotfiles" that may have modified PATH unconditionally (e.g. on any
+# host running nix-darwin(!)).
+#
+# Replacing it with a test that checks that a) activation puts the FLOX_ENV/bin
+# first, and b) that it doesn't put it there more than once.
```
```suggestion
# The following tests check that repeat in-place activations, such as from RC
# files, modify PATH to:
#
# 1. put FLOX_ENV/bin first
# 2. not put it there more than once
#
# This guards against:
#
# 1. RC files unconditionally modifying PATH (e.g. nix-darwin)
# 2. `path_helper` on macOS putting our environment last
```

dcarley commented on assets/activation-scripts/activate:
> To discuss: relation of `FLOX_ENV_DIRS` vs. `_FLOX_ACTIVE_ENVIRONMENTS`, and which is the source of truth, particularly when mixing rust and activate script activations (e.g. as happens in the nested environments found with a build). At the moment I'm finding that all tests pass when `_FLOX_ACTIVE_ENVIRONMENTS` does _not_ match reality, and wondering if that means our test coverage is missing something or if we shouldn't a) consolidate these into a single variable, managed from the `flox-env-helper` script.

I'm not quite sure that I follow this. Can you show the setup and effect in a test or manual reproduction?

limeytexan replied:
Example: for builds we perform the following nested activation:
```
          $(FLOX_INTERPRETER)/activate --env $(FLOX_ENV) --mode dev --turbo -- \                                                                                                                                                               
            $(_env) -i out=$(_out) $(foreach i,$(ALLOW_OUTER_ENV_VARS),$(i)="$$$$$(i)") \                                                                                                                                                      
              $(_build_wrapper_env)/activate --env $(_build_wrapper_env) --mode dev --turbo -- \                                                                                                                                               
                $(_t3) $($(_pvarname)_logfile) -- $(_bash) -e $($(_pvarname)_buildScript)                                                                                                                                                      
```

This doesn't involve the `flox` CLI at all, so therefore `_FLOX_ACTIVE_ENVIRONMENTS` is not set for the subsequent build, but `FLOX_ENV_DIRS` (and `FLOX_ENV_LIB_DIRS`) are set, and it works fine.

But while the above example demonstrates two nested script-based activations, this can also be invoked from within an interactive `flox activate` for the project, and that can often take place within the dotfile-sourced `eval "$(flox activate)"` invocation performed for the default environment. So the bottom line is that by the time that build script is invoked:
* `_FLOX_ACTIVE_ENVIRONMENTS` can reflect up to two environments
* `FLOX_ENV_DIRS` will reflect those, plus any additional ones invoked within that shell's own activation "stack"
* unsure whether those script-based activations will be "off radar" from the flox CLI's perspective (for the purpose of subsequent "attach" activations)?!?

ysndr replied:
_FLOX_ACTIVE_ENVIRONMENTS

Is currently only used by the cli
1. To avoid reactivation
2. To list active environments in `flox envs`
3. I think the envs in the prompt are also derived from this but I'm not sure if that has changed.

Besides, _FLOX_ACTIVE_ENVIRONMENTS is also not directly replaceable with FLOX_ENV_DIRS as it only contains environment metadata

I.e even before this or neither as the source of truth really

limeytexan replied:
I agree that this isn't a new issue and not urgent to address as part of this PR, but I imagine would be good to discuss further and consider with the next bit of work we perform in this space. For now I'll click resolve - thanks for your thoughts!

dcarley commented on assets/activation-scripts/activate.d/flox-env-helper.awk:
```diff
@@ -0,0 +1,194 @@
+
+# No input block means script will exit immediately after BEGIN block.
+
+# Function definitions follow.
```
(non-blocking, review commentary) I didn't review the logic of the functions too carefully. Relying mostly on test coverage. I did wonder if we should test this script in isolation but I think it's better to integration test the effects on activation.

limeytexan replied:
Thanks for the reminder, it would be trivial to test this in isolation, e.g.:
```
flox [limeytexan/default] Michaels-MacBook-Pro% FLOX_ENV_DIRS=foo/bar:baz/bing FLOX_ENV=bang/bong awk -f assets/activation-scripts/activate.d/flox-env-helper.awk bash
export FLOX_ENV_DIRS="bang/bong:foo/bar:baz/bing";
export FLOX_ENV_LIB_DIRS="bang/bong/lib:foo/bar/lib:baz/bing/lib";
export PATH="bang/bong/bin:bang/bong/sbin:foo/bar/bin:foo/bar/sbin:baz/bing/bin:baz/bing/sbin:/Users/brantley/.flox/run/aarch64-darwin.default.run/bin:/Users/brantley/.flox/run/aarch64-darwin.default.run/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin";
export MANPATH="bang/bong/share/man:foo/bar/share/man:baz/bing/share/man:/Users/brantley/.flox/run/aarch64-darwin.default.run/share/man:/usr/share/man:/usr/local/share/man:";
flox [limeytexan/default] Michaels-MacBook-Pro% 
```

But agree the integration tests already provide the full coverage we need.

limeytexan commented on assets/activation-scripts/activate.d/flox-env-helper.awk:
```diff
@@ -0,0 +1,172 @@
+    print "export PATH=\"" new_path "\";"
+    print "export MANPATH=\"" new_manpath "\";"
+  } else {
+    print("Unknown shell: " flox_shell) > "/dev/stderr"
```
Not sure what you mean as any utility should direct error messages to stderr. It's certainly possible that the program invoking the script might discard, hide or otherwise disregard the warning but at least it will be visible in the output of `strace` for those trying to figure out what's going on.

Let me know if that doesn't make sense, and feel free to reply to reopen the conversation.

ysndr replied:
Tangential, but we tend to use `>&2` elsewhere when we echo to stderr.

Not saying it's big deal but something I note on the basis of consistency 🤷🏼‍♂️

dcarley replied:
I vaguely recalled that there are some gotchas about using `/dev/stderr` in some situations because it's either:

1. A symlink to FD2 of the current process which can cause problems with `sudo` or containers:
  a. https://unix.stackexchange.com/a/38580
  b. https://superuser.com/a/1699427
  c. https://community.fly.io/t/container-non-root-user-cant-write-to-dev-stdout-or-dev-stderr/10375

2. Interpreted by tools in their own way:
  a. https://www.gnu.org/software/gawk/manual/html_node/Special-FD.html

I'm not sure we'd know if it didn't work as expected though. Come to think of it, it's now on the hot path for determining supported shells, which we should probably add tests for and potentially move our existing check higher: https://github.com/flox/flox/blob/60e77f4da90800cdb706075b3c7ccfcf9d87766b/assets/activation-scripts/activate#L176-L186

limeytexan replied:
> Tangential, but we tend to use >&2 elsewhere when we echo to stderr.

Sadly that doesn't work in the awk language. (Assuming you're highlighting that vs. the distinction between `>&2` and `1>&2` which is the same thing on account of FD 1 being the default for redirection.)

> I vaguely recalled that there are some gotchas about using /dev/stderr in some situations ...

My recollection of course goes back to any of a dozen UNIX variants that don't have it in the first place, but as we're only supporting Linux and Darwin I think we're on solid ground. I do wish that awk had a native way of redirecting to stderr but alas.  🤷

limeytexan replied:
I honestly expect we'll be fine using `/dev/stderr`, but the only other viable option I've found is to use something like `system("echo foo >&2")`.

I'll go ahead and click resolve for now but please do reply to reopen the conversation if y'all would prefer I change as above.

limeytexan replied:
Ah ... just found out that with `gawk` accesses to `/dev/stderr` are actually implemented with an internal file descriptor rather than using the filesystem:
```
 Special File Names
    When  doing I/O redirection from either print or printf into a file,
    or via getline from a file, gawk recognizes certain special filenames
    internally. These filenames allow access to open file descriptors
    inherited from gawk's parent process (usually the shell). These file
    names may also be used on the command line to name data files.
    The filenames are:

    -           The standard input.
    /dev/stdin  The standard input.
    /dev/stdout The standard output.
    /dev/stderr The standard error output.
    /dev/fd/n   The file associated with the open file descriptor n.
```
 So another easy fix here would be to move from `nawk` to `gawk`.

dcarley replied:
Yeah, that link above says that some other AWK implementations do it to, but I can't find anything about `nawk`.

I think it's probably fine to leave as-is but it would be good to address the test and move the existing check I mentioned.

limeytexan replied:
Spoke with Dan and confirmed the check of `$FLOX_SHELL` is OK where it is, because we're not referring to it prior to that point. For the avoidance of doubt, the invocation of `flox-env-helper` that does appear a few lines prior to line 176 is being called with the explicit arg of "bash" because it's being sourced from the [bash] `activate` script.

limeytexan commented on assets/activation-scripts/activate.d/flox-env-helper.awk:
```diff
@@ -127,9 +149,9 @@ function basename(file, a, n)
+function prepend_if_not_found(insert, path, path_array, seen_hash, i, result)
 {
-  # local variables: path_array, seen_hash, i
+  # local variables: path_array, seen_hash, i, result
```
Yeah .. I didn't know that until writing this either!

> It is conventional to place some extra space between the arguments and the local variables, in order to document how your function is supposed to be used.

Yeah .. I did that, but then `gawk --pretty-print` put them back on the same line, so I instead opted to document the local variables in a comment below. It doesn't much matter but happy to reformat according to group consensus! Please reply or commit a patch if you want to change it!

ysndr replied:
Awkward

Question: Why do we even use ask for this?
Guess it's the closest thing to "light scripting language that isn't too bonkers and knows what strings are" (although I'm not too sure about the latter two statements)

ysndr replied:
Did the pretty printer also complain about prefixing them?

limeytexan replied:
> Question: Why do we even use ask for this?

Extremely quick to create, easily understood/modified, very convenient builtins for PATH parsing/modification, lightest possible closure impact (as nawk is already in the closure), and also recall that this replaces logic previously written as an awk one-liner embedded within a shell script (which was much uglier). It could be rewritten in anything, and we may one day do just that, but for now this gets the job done with high quality, best performance, smallest size, in the least amount of time.

limeytexan replied:
> Did the pretty printer also complain about prefixing them?

Prefixing what exactly? Do you mean local variables starting with the `_` character? If so then no I didn't try that. I wouldn't call `gawk --pretty-print` a linter, it's more of a reformatter.

limeytexan commented on cli/tests/activate.bats:
Agree. In this case it's that `BADPATH` introduced by the dotfiles that prevents the `PATH` from leaking through from the flox-cli-tests shell, and correspondingly `jq` is not found. It was definitely not obvious and somewhat hard to spot. Agree this should be split out into a separate effort.

limeytexan commented on assets/activation-scripts/activate.d/generate-bash-startup-commands.bash:
```diff
@@ -42,23 +40,21 @@ generate_bash_startup_commands() {
   # Set the prompt if we're in an interactive shell.
   echo "if [ -t 1 ]; then source '$_activate_d/set-prompt.bash'; fi;"
 
+  # We already customized the PATH and MANPATH, but the user and system
+  # dotfiles may have changed them, so finish by doing this again.
```
Yes, it's referring to exactly that, the invocation that takes place near the top of the `activate` script. :+1:

ysndr replied:
Why do we need both?

limeytexan commented on cli/tests/activate.bats:
```diff
@@ -1505,6 +1505,59 @@ EOF
+# first, and b) that it doesn't put it there more than once.
+
+# bats test_tags=activate,activate:inplace-reactivate,activate:inplace-reactivate:bash
+@test "bash: 'flox activate' patches PATH correctly when already activated" {
```
Possibly on account of the quoting issues that Matthew described in commit 6128bbb?

Also, I force-pushed a couple of updates to that patch and would just want to confirm that you tested with the one that looks like this?
```
  FLOX_SHELL="bash" run -- \
    "$FLOX_BIN" activate -- \
    $_bash -c "source <($FLOX_BIN activate); $_bash $TESTS_DIR/activate/verify_PATH.bash"
```
Notably, there is no variable expansion performed in this version throughout the bats -> flox -> bash -> flox -> bash invocation chain after that first replacement of `$FLOX_BIN` and `$TESTS_DIR`.

dcarley replied:
It was a `git cherry-pick 9e66ca6ff6a512a61b86b14cf93a1f2af59a5e78` with no conflicts, so it matched what was on this PR: https://github.com/flox/flox/commit/#diff-7d1ec57a118c271f57548d87b2997a82d9e036adba0c5e6facf39f35ff7bf672R1520-R1522

limeytexan replied:
I'll admit I'm not following what you mean above - I clicked on the PR and browsed files at that commit but didn't see the `activate:inplace-reactivate` test in there? Regardless, if it's OK I'll go ahead and resolve this for the sake of moving forward - please reply if you want us to do anything specific on this prior to merging.

ysndr commented on assets/activation-scripts/activate.d/flox-env-helper.awk:
```diff
@@ -0,0 +1,194 @@
+  return result
+}
+
+function remove_path_dups(path, path_array, dedup_array, seen_hash, i, dedup_count, result)
```
Nit: I'd still prefer marking local variables somehow e.g. `_var` just an inline comment is easily outdated or overlooked.

limeytexan replied:
Ah, OK. I agree for all but `i` as that's just a universal counter. Will update accordingly.  👍

limeytexan replied:
Renamed local variables in commit 7da7cc86ebad225674de34a3bbe075b70ad6fb77

ysndr commented on assets/activation-scripts/activate.d/flox-env-helper.awk:
```diff
@@ -0,0 +1,194 @@
+    _prepend_path = _prepend_path flox_env_dirs[i] "/bin:" flox_env_dirs[i] "/sbin:"
+    _prepend_manpath = _prepend_manpath flox_env_dirs[i] "/share/man:"
+  }
+  _prepend_path = _prepend_path ENVIRON["FLOX_ENV"] "/bin:" ENVIRON["FLOX_ENV"] "/sbin:"
```
Can we document the env variables this depends on in the docs above?
It seems if this script is run without FLOX_ENVS it won't work as expected (or at all?).

Currently that's always the case in practice but it's a weak dependency that we should at least acknowledge

ysndr replied:
Yes I think a check would be good to make that an explicit invariant

limeytexan replied:
Added a check in commit 1971aea2fc110340397f1dcca64c9fc7acd28ca4 ... resolving conversation.

ysndr commented on cli/tests/activate.bats:
```diff
@@ -2624,9 +2680,10 @@ EOF
   project_setup
   run bash <(cat <<'EOF'
+    _expect="$(command -v expect)"
     eval "$("$FLOX_BIN" activate)"
 
-    expect "$TESTS_DIR/activate/activate.exp" "$PROJECT_DIR"
+    $_expect "$TESTS_DIR/activate/activate.exp" "$PROJECT_DIR"
```
question: What's wrong with expect from the PATH?

dcarley replied:
It gets removed from `PATH` after the in-place activation before this line. Related to:

- a37bf1598e6013d17473c63bc70040df8dde687f
- https://github.com/flox/flox/pull/2552#discussion_r1900758571

limeytexan replied:
Thanks Dan, resolving conversation.

limeytexan commented on assets/activation-scripts/activate:
```diff
@@ -156,6 +156,10 @@ fi
 _FLOX_ENV="$($_dirname -- "${BASH_SOURCE[0]}")"
```
If you look at the next line its honoring a predefined `FLOX_ENV` and just using the [derived] `_FLOX_ENV` as a fallback value. So yes, it is only ever used if the CLI hasn't already defined the canonical path for the environment.  👍

limeytexan replied:
I'll go ahead and resolve this for now - please reply to reopen if you think necessary.

limeytexan commented:
FTR [Eric confirmed that this PR addresses the problems he was seeing with 1.3.8](https://discourse.flox.dev/t/path-issue-with-1-3-8-on-mac-zsh-direnv/1023/3?u=brantley).

mkenigs commented on assets/activation-scripts/activate.d/zdotdir/.zprofile:
```diff
@@ -23,11 +23,12 @@ restore_saved_vars() {
-    export _FLOX_RESTORE_PATH="$_save_FLOX_RESTORE_PATH"
-    export _FLOX_RESTORE_MANPATH="$_save_FLOX_RESTORE_MANPATH"
     export _FLOX_ACTIVATION_PROFILE_ONLY="$_save_FLOX_ACTIVATION_PROFILE_ONLY"
+    source =("$_flox_env_helper" "zsh")
```
Looks like we're running _flox_env_helper 10 times for a single activation. That seems like a lot. Can we just do it twice like we are for other shells?

mkenigs replied:
The reasoning is to defend against breakage of PATH or MANPATH from any file in ZDOTDIR. If we can refactor this to be just twice and it's passing tests, that may be okay. We need to make sure that before we run anything that depends on PATH or MANPATH ourselves, we do cleanup

mkenigs commented on assets/activation-scripts/activate.d/generate-fish-startup-commands.bash:
```diff
@@ -53,6 +43,10 @@ generate_fish_startup_commands() {
 
+  # We already customized the PATH and MANPATH, but the user and system
+  # dotfiles may have changed them, so finish by doing this again.
+  echo "$_flox_env_helper fish | source;"
```
Per the pattern in this file, FLOX_ENV should be passed as an argument to `_flox_env_helper` rather than being passed implicitly via the environment

mkenigs replied:
I think this line can be `$_flox_env_helper fish` instead of `echo "$_flox_env_helper fish | source;"`

mkenigs commented on assets/activation-scripts/activate.d/flox-env-helper.awk:
```diff
@@ -0,0 +1,199 @@
+#!@nawk@/bin/nawk -f
```
This script would be a good candidate for a `flox-activations` subcommand.

zmitchell replied:
Big +1

zmitchell replied:
We get antsy when we want to add a new language, and recently we've added Perl for the environment builder, Awk for this script, Make (shudder) for the package builder, and C for `t3`. I also already have this logic in the Rust experiment I did.

mkenigs commented on assets/activation-scripts/activate.d/flox-env-helper.awk:
```diff
@@ -0,0 +1,199 @@
+  # Emit commands to set variables for the various shell dialects.
+  flox_shell_basename = basename(ARGV[1])
+  if (flox_shell_basename == "bash") {
+    print "export FLOX_ENV_DIRS=\"" new_flox_env_dirs "\";"
+    print "export FLOX_ENV_LIB_DIRS=\"" new_flox_env_lib_dirs "\";"
```
If the CLI isn't setting these variables to incorrect values, we could move this from the helper to the activate script. We'd also want to verify that an in-place activation from an RC file is setting these correctly - i.e. `flox activate -d profile` then implicit `eval "$(flox activate -d default)"` should make sure that `profile` is before default. At least if the default env was already active
